### ISSUE TITLE
Webform-Specific E-Mail 'from' settings not overriding site defaults to set the Reply-to address properly

### ISSUE THREAD

**Issue by bobchristenson** [backdrop-contrib/webform#157]
We've got webform module (current version 1.x-4.21.0) setup with defaults in the config screen...so a 'sitewide' email address and name that is used for E-Mail sending.

Then, we've got a webform setup.  In the e-mail configs for that specific form we've overridden the 'from' email address so it should appear to send from the "Email" component and the name from the webform name component (submitters email and name).

When the email arrives, however, it is coming from the default configurations email address and name.

It seems like the specific form's settings isn't overriding the defaults.  Has anyone seen this problem?  Any thoughts on a fix?

**Comment by bobchristenson** [comment]
Update, actually...it seems it's not even using the webform defaults.  I think it's using the general site email address and then using the first part (before the @) as the name...

**Comment by bobchristenson** [comment]
OK, looks like it was this same issue: https://www.drupal.org/node/461324
I ended up, to at least get a 'reply-to' address functional, doing this:

```
function mymodule_mail_alter(&$message) {
	if($message['id'] == 'webform_submission' && $message['module'] == 'webform'){
		$from_info = $message['params']['email']['reply-to'];
		$message['headers']['Reply-To'] = $from_info;
	}
}
```

I tried to rewrite some of the 'sender' info to make it look like it was coming from the submitters name/email but had no success...it seemed to always be rewritten or otherwise blocked by the mailserver if I did that.

As mentioned in the Drupal issue, it's actually Backdrop core that overrides anything set in the Webform email settings.  For a while I thought this info might be rewritten by the server configs or the receiver's email service...but it doesn't appear to be the case.  Which means this is ALWAYS happening?

If so...what's the point of the individual Webform email configs (designating 'from' name/email) at all if they'll always be overwritten?

**Comment by indigoxela** [comment]
Took some time here to dig a bit deeper. This looks like a bug, but might _eventually_ only give that impression because of bad wording of form labels and help texts. Or maybe there's really a bug.

First of all: in general it's never a good idea to use a different email domain (not yours) as sender. SPF records might completely prevent these mails in worst case. That's why Webform chose to provide a Reply-To header setting some time ago.

Under admin/config/content/webform there's a possibility to set _default_ Webform email values, including "From address". Besides the naming thing there, this setting works as expected, I guess.

Under individual Webform settings (node/XX/webform/emails/X) there's also a form item "E-mail from address".
And that setting is completely ignored if the default setting has "Use Reply-To header" checked. Otherwise (unchecked) the chosen "From" is actually used as _Reply-To_ ... WTF?

So we have:

1. confusing form labels and descriptions
2. confusing logic behind the per-webform setting

This causes way too much head scratching. Could we please simplify that?
The first question (which I can't answer) is: how is this actually supposed to be?

**Comment by ghost** [comment]
I'm also experiencing this. I believe it's a bug as one of my clients only recently noticed it happening (i.e. I'm pretty sure it used to work as expected). Need to spend some more time looking into it properly though...

**Comment by ghost** [comment]
Here's a PR that works for me: https://github.com/backdrop-contrib/webform/pull/174

**Comment by herbdool** [comment]
The following situation seems to work as I expect: Reply-To checked.

![Screenshot from 2021-01-04 13-41-14](https://user-images.githubusercontent.com/1071483/103568638-ea1f2200-4e93-11eb-9f48-060f26c31158.png)

Results in:

![Screenshot from 2021-01-04 15-38-16](https://user-images.githubusercontent.com/1071483/103577479-e8108f80-4ea2-11eb-9cec-155ed8093c11.png)

The From is the site wide email and the Reply-To is what is set in the Webform settings. That seems right to me. What I don't understand from the UI is what is supposed to happen if Reply-To is unchecked. The description isn't a proper sentence so it's vague: "If the default from name and address are set above, send all e-mail from the default address set the "Reply-To" header to the actual sender."

Looks like I need to compare with a D7 site.

**Comment by herbdool** [comment]
Here's the d7 equivalent:
![d7-webform-settings](https://user-images.githubusercontent.com/1071483/103576095-9c5ce680-4ea0-11eb-934c-0a8a2ba21bf8.png)

And the headers:
![Screenshot from 2021-01-04 15-39-16](https://user-images.githubusercontent.com/1071483/103577583-18f0c480-4ea3-11eb-8472-ea6cad039a82.png)

So the Webform email default shouldn't override the individual webform node settings for the Reply-to and the From is the Webform email default. I'll see if the PR above helps with this.

Update: @BWPanda, I tried the PR but doesn't seem to change anything.

**Comment by ghost** [comment]
> That seems right to me.

@herbdool That seems right to me too, but I don't know how you got that result... Here's what I did:

Installed Webform from git (i.e. 1.x-4.x), left all settings as default (i.e. Reply-To header checked), and created a webform with email address, name and message fields. I setup an email to be sent to me on submission, with the default subject, but with the Email component as the Email From Address, and the Name component as the Email From Name.

![image](https://user-images.githubusercontent.com/2385329/103592060-14c9b300-4f46-11eb-913e-ca9e7ce9fb54.png)

I then submitted the webform with the following information:

Email: bob@example.com
Name: Bob
Message: Hi there!

And got the following email headers:

```
Reply-To: "Bob via Sandbox" <noreply@lndo.site>
Return-Path: <noreply@lndo.site>
Sender: noreply@lndo.site
Subject: Form submission from: Test
```

The Reply-To address is wrong. Nowhere in the headers does `bob@example.com` appear.

I then switched to my PR branch, cleared caches and submitted the webform again:

Email: mary@example.com
Name: Mary
Message: Anybody home?

And got the following email headers:

```
Reply-To: "Mary" <mary@example.com>
Return-Path: <noreply@lndo.site>
Sender: noreply@lndo.site
Subject: Form submission from: Test
```

Now it's correct.

**Comment by herbdool** [comment]
@BWPanda your result doesn't match D7's either. What is in the "From"?

**Comment by ghost** [comment]
@herbdool From, in both of the above cases I described, was `noreply@lndo.site`.

And to me that makes sense. Emails are being sent to you from the website, period. You can't change that. It's not as if the people are emailing you from their accounts directly. So that's fine having the From come from the site address.

Then with the patch applied and the Reply-To box checked, the email address you send to when you hit reply in the email goes to the person, rather than the website. That also makes sense to me.

**Comment by herbdool** [comment]
Here are samples from recent emails to me for comparison:

```
From: CXXXX XXXXXXXX - Pantheon Manager of Technical Documentation <CSEInfo@pantheon.io>
Reply-To: CSEInfo@pantheon.io
```
and

```
From: "Joe Schmoe (Basecamp)" <notifications@basecamp.com>
Reply-To: document-XXXXXXX-XXXXXXXXXXXXXXXX@basecamp.com
```

Compare those with my example from Drupal 7 Webform:

```
'From' => '"admin via d7 webform" <noreply-webform@example.com',
'Reply-To' => '"admin" <admin@example>',
```

**Comment by ghost** [comment]
@herbdool I'm not sure what point you're making... Are you saying Drupal is doing it right, or wrong?

**Comment by herbdool** [comment]
Wanted to first find some examples and save them here before I have a chance to get back to this. Hopefully soon.

**Comment by herbdool** [comment]
The main oddness here is because of the way `backdrop_mail` (and `drupal_mail` in D8) changed. See: https://github.com/backdrop/backdrop-issues/issues/305. The `$from` ended up being fixed to the site domain with no way to alter it. There's an equivalent issue for D8 (where it was merged https://www.drupal.org/node/2164905) and D7 (where it hasn't been merged as of writing this): https://www.drupal.org/node/111702.

Since *all* emails are now being set with the Reply-to set, there's not much point to the "Use Reply-to header" option in Webform. It's going to be set either way so we'll always format in Webform. So instead I'm working on a PR to remove it, update the wording of the other options. And also in `includes/webform.submissions.inc` we should pass `$email['reply-to']` instead of `$email['from']` to reflect the change in `backdrop_mail`.

Per webform settings:

![Screenshot from 2021-08-12 14-59-07](https://user-images.githubusercontent.com/1071483/129254334-ba013c1d-1a1d-4867-a374-5eaf65b218e3.png)

Global Webform settings [Updated: I've changed it back to "From" and added to description to say that this should match the sending domain.]:

![Screenshot from 2021-08-13 09-34-07](https://user-images.githubusercontent.com/1071483/129367882-1ba5fbf2-ae66-4ec1-9f8f-c5accd4b2540.png)


There's still an issue where it's not possible to get the From display name to show something like `"Billy the Kid via Default Sender Name" <noreply@lndo.site>`. Instead we get:

```
From: noreply@lndo.site
...
Reply-To: "Billy the Kid" <billy@robthattrain.com>
Return-Path: <noreply@lndo.site>
Sender: noreply@lndo.site
```

We can use `hook_mail` to set the From address as what is set in the global Webform settings. 

[Updated: I've now added this to `hook_mail`. And this is how it looks:]

```
From: "Billy the kid via Default Sender Name" <default_sender@webform.org>
...
Reply-To: "Billy the kid" <billy@bailmeout.com>
Return-Path: <noreply@lndo.site>
Sender: noreply@lndo.site
```

(The Sender and Return-Path might not match the domain of the From, but I believe that's how it works in D7 as well. An SMTP service might override this, and this is not something to fix in Webform.)

**Comment by herbdool** [comment]
Here's my take on fixing the Reply-to address: https://github.com/backdrop-contrib/webform/pull/185. Since core *always* sets the reply-to header we can just always configure it. In my PR it reliably will first check the individual webform settings and use the global Webform settings if set to default.

The "from" address can be a different issue.

**Comment by herbdool** [comment]
Sorry, I'm going to make a change by reverting the global Webform settings back to saying "From" and just put a warning that the domain should probably match the website or their mailing service.

**Comment by herbdool** [comment]
Alright I think the PR is working how I want it to work. I've addressed it in a way so it closely as possible matches the way D7 Webform with Reply-to header enabled will work. I believe @indigoxela has the right idea here. See my comment above https://github.com/backdrop-contrib/webform/issues/157#issuecomment-897902034 for how this PR now looks (I've updated the comment). @indigoxela's comment above https://github.com/backdrop-contrib/webform/issues/157#issuecomment-640486049:

> First of all: in general it's never a good idea to use a different email domain (not yours) as sender. SPF records might completely prevent these mails in worst case. That's why Webform chose to provide a Reply-To header setting some time ago.

I agree. And since core now *always* sets a Reply-to header I won't try to override that in Webform. Instead we can remove the option in Webform since we can't turn it off.

> Under admin/config/content/webform there's a possibility to set default Webform email values, including "From address". Besides the naming thing there, this setting works as expected, I guess.

I've updated the description here so perhaps it's clearer that this address needs to match the sending domain.

> Under individual Webform settings (node/XX/webform/emails/X) there's also a form item "E-mail from address".
And that setting is completely ignored if the default setting has "Use Reply-To header" checked. Otherwise (unchecked) the chosen "From" is actually used as Reply-To ... WTF?

This, I believe, I've fixed by changing this line: 

```diff
- $message = backdrop_mail('webform', 'submission', $address, $language, $mail_params, $email['from']);
+ $message = backdrop_mail('webform', 'submission', $address, $language, $mail_params, $email['reply-to']);
```
and updating the labels from "From" to "Reply-to" for those. They should only appear in the Reply-to header.

**Comment by indigoxela** [comment]
I think, it's OK. I tested locally, added two email settings to a webform, these are the outcoming mail header values:

From site to component value:
```
Return-Path: <noreply@bdroptesting.mydomain.tld>
To: notifyme@mydomain.tld
Sender: noreply@bdroptesting.mydomain.tld
From: "bdroptesting.mydomain.tld via bdroptesting.mydomain.tld" <noreply@bdroptesting.mydomain.tld>
Reply-To: "bdroptesting.mydomain.tld" <noreply@bdroptesting.mydomain.tld>
```

Component value as replyto:
```
Return-Path: <noreply@bdroptesting.mydomain.tld>
To: test@mydomain.tld
Sender: noreply@bdroptesting.mydomain.tld
From: "bdroptesting.mydomain.tld via bdroptesting.mydomain.tld" <noreply@bdroptesting.mydomain.tld>
Reply-To: "bdroptesting.mydomain.tld" <notifyme@mydomain.tld>
```
The odd "via" thing isn't optimal (comes from includes/webform.submissions.inc `$from_name = t('!name via !site_name',`), but that could also be a follow-up.

**Comment by herbdool** [comment]
Thanks @indigoxela. I think it's ready then.

I'm not sure what you mean about the "via" so sounds like a new issue. This PR didn't change the way that works.
